# This is the main notebook file for the machine learning program

In [1]:
# Imports
import pandas as pd
from pathlib import Path
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.preprocessing import StandardScaler,OneHotEncoder
from sklearn.metrics import accuracy_score
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
%run read_csv_files.ipynb

2023-02-14 11:41:45.335033: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


## Begin with scaled data where the target value predicts whether the home team will win the game (Signal = '1')
## Part 1: Neural Network with Deep Learning for 10 Seasons

In [2]:
# Drop the 'Date' column from the dataframe
ten_seasons_nn = df_combined_2011_2021.drop(columns=['Date'])

# Display the dataframe
ten_seasons_nn

Home_Team             Away_Team  Over_Under  Spread  \
0            New York Knicks        Boston Celtics       195.5     5.0   
1           Dallas Mavericks            Miami Heat       190.5     3.5   
2         Los Angeles Lakers         Chicago Bulls       186.0     1.5   
3      Oklahoma City Thunder         Orlando Magic       197.0     7.5   
4      Golden State Warriors  Los Angeles Clippers       212.0     4.0   
...                      ...                   ...         ...     ...   
12576           Phoenix Suns       Milwaukee Bucks       221.0     5.5   
12577        Milwaukee Bucks          Phoenix Suns       222.0     3.5   
12578        Milwaukee Bucks          Phoenix Suns       220.5     3.5   
12579           Phoenix Suns       Milwaukee Bucks       218.5     4.0   
12580        Milwaukee Bucks          Phoenix Suns       223.0     5.0   

       Moneyline_Home_Team  Moneyline_Away_Team  Game_Total_Points  Win_Margin  
0                   -210.0                185.0              210.0         2.0  
1                    185.0               -210.0              199.0       -11.0  
2                    205.0               -245.0              175.0        -1.0  
3                   -380.0                310.0              186.0         8.0  
4                    175.0               -200.0              191.0       -19.0  
...                    ...                  ...                ...         ...  
12576               -180.0                160.0              226.0        10.0  
12577               -210.0                180.0              220.0        20.0  
12578               -190.0                170.0              212.0         6.0  
12579               -170.0                150.0              242.0        -4.0  
12580               -190.0                170.0              203.0         7.0  

[12581 rows x 8 columns]

In [3]:
# Review the data types
ten_seasons_nn.dtypes

Home_Team               object
Away_Team               object
Over_Under             float64
Spread                 float64
Moneyline_Home_Team    float64
Moneyline_Away_Team    float64
Game_Total_Points      float64
Win_Margin             float64
dtype: object

In [4]:
# Create a list of the categorical variables
categorical_variables_1 = list(
    ten_seasons_nn.dtypes[ten_seasons_nn.dtypes == 'object'].index
)

# Display the list
categorical_variables_1

['Home_Team', 'Away_Team']

In [5]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(
    sparse = False
)

In [6]:
# Encode the categorical variables into numerical format
encoded_data_1 = enc.fit_transform(
    ten_seasons_nn[categorical_variables_1]
)

In [7]:
# Create a dataframe with the encoded data
encoded_ten_seasons_df = pd.DataFrame(
    encoded_data_1,
    columns = enc.get_feature_names_out(
        categorical_variables_1
    )
)

# Display the dataframe
encoded_ten_seasons_df

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  \
0                          0.0                       0.0   
1                          0.0                       0.0   
2                          0.0                       0.0   
3                          0.0                       0.0   
4                          0.0                       0.0   
...                        ...                       ...   
12576                      0.0                       0.0   
12577                      0.0                       0.0   
12578                      0.0                       0.0   
12579                      0.0                       0.0   
12580                      0.0                       0.0   

       Home_Team_Brooklyn Nets  Home_Team_Charlotte Bobcats  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   
...                        ...                          ...   
12576                      0.0                          0.0   
12577                      0.0                          0.0   
12578                      0.0                          0.0   
12579                      0.0                          0.0   
12580                      0.0                          0.0   

       Home_Team_Chicago Bulls  Home_Team_Cleveland Cavaliers  \
0                          0.0                            0.0   
1                          0.0                            0.0   
2                          0.0                            0.0   
3                          0.0                            0.0   
4                          0.0                            0.0   
...                        ...                            ...   
12576                      0.0                            0.0   
12577                      0.0                            0.0   
12578                      0.0                            0.0   
12579                      0.0                            0.0   
12580                      0.0                            0.0   

       Home_Team_Dallas Mavericks  Home_Team_Denver Nuggets  \
0                             0.0                       0.0   
1                             1.0                       0.0   
2                             0.0                       0.0   
3                             0.0                       0.0   
4                             0.0                       0.0   
...                           ...                       ...   
12576                         0.0                       0.0   
12577                         0.0                       0.0   
12578                         0.0                       0.0   
12579                         0.0                       0.0   
12580                         0.0                       0.0   

       Home_Team_Detroit Pistons  Home_Team_Golden State Warriors  ...  \
0                            0.0                              0.0  ...   
1                            0.0                              0.0  ...   
2                            0.0                              0.0  ...   
3                            0.0                              0.0  ...   
4                            0.0                              1.0  ...   
...                          ...                              ...  ...   
12576                        0.0                              0.0  ...   
12577                        0.0                              0.0  ...   
12578                        0.0                              0.0  ...   
12579                        0.0                              0.0  ...   
12580                        0.0                              0.0  ...   

       Away_Team_Oklahoma City Thunder  Away_Team_Orlando Magic  \
0                                  

In [8]:
# Create a new variable for the exisiting numerical data and reset the index
numerical_variables_1 = ten_seasons_nn.drop(
    columns = categorical_variables_1
)
numerical_variables_1

Over_Under  Spread  Moneyline_Home_Team  Moneyline_Away_Team  \
0           195.5     5.0               -210.0                185.0   
1           190.5     3.5                185.0               -210.0   
2           186.0     1.5                205.0               -245.0   
3           197.0     7.5               -380.0                310.0   
4           212.0     4.0                175.0               -200.0   
...           ...     ...                  ...                  ...   
12576       221.0     5.5               -180.0                160.0   
12577       222.0     3.5               -210.0                180.0   
12578       220.5     3.5               -190.0                170.0   
12579       218.5     4.0               -170.0                150.0   
12580       223.0     5.0               -190.0                170.0   

       Game_Total_Points  Win_Margin  
0                  210.0         2.0  
1                  199.0       -11.0  
2                  175.0        -1.0  
3                  186.0         8.0  
4                  191.0       -19.0  
...                  ...         ...  
12576              226.0        10.0  
12577              220.0        20.0  
12578              212.0         6.0  
12579              242.0        -4.0  
12580              203.0         7.0  

[12581 rows x 6 columns]

In [9]:
# Add the numerical data and the categorical encoded data to one dataframe
encoded_ten_seasons_df = pd.concat(
    [encoded_ten_seasons_df,
     numerical_variables_1],
    axis=1
)

# Display the dataframe
encoded_ten_seasons_df

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  \
0                          0.0                       0.0   
1                          0.0                       0.0   
2                          0.0                       0.0   
3                          0.0                       0.0   
4                          0.0                       0.0   
...                        ...                       ...   
12576                      0.0                       0.0   
12577                      0.0                       0.0   
12578                      0.0                       0.0   
12579                      0.0                       0.0   
12580                      0.0                       0.0   

       Home_Team_Brooklyn Nets  Home_Team_Charlotte Bobcats  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   
...                        ...                          ...   
12576                      0.0                          0.0   
12577                      0.0                          0.0   
12578                      0.0                          0.0   
12579                      0.0                          0.0   
12580                      0.0                          0.0   

       Home_Team_Chicago Bulls  Home_Team_Cleveland Cavaliers  \
0                          0.0                            0.0   
1                          0.0                            0.0   
2                          0.0                            0.0   
3                          0.0                            0.0   
4                          0.0                            0.0   
...                        ...                            ...   
12576                      0.0                            0.0   
12577                      0.0                            0.0   
12578                      0.0                            0.0   
12579                      0.0                            0.0   
12580                      0.0                            0.0   

       Home_Team_Dallas Mavericks  Home_Team_Denver Nuggets  \
0                             0.0                       0.0   
1                             1.0                       0.0   
2                             0.0                       0.0   
3                             0.0                       0.0   
4                             0.0                       0.0   
...                           ...                       ...   
12576                         0.0                       0.0   
12577                         0.0                       0.0   
12578                         0.0                       0.0   
12579                         0.0                       0.0   
12580                         0.0                       0.0   

       Home_Team_Detroit Pistons  Home_Team_Golden State Warriors  ...  \
0                            0.0                              0.0  ...   
1                            0.0                              0.0  ...   
2                            0.0                              0.0  ...   
3                            0.0                              0.0  ...   
4                            0.0                              1.0  ...   
...                          ...                              ...  ...   
12576                        0.0                              0.0  ...   
12577                        0.0                              0.0  ...   
12578                        0.0                              0.0  ...   
12579                        0.0                              0.0  ...   
12580                        0.0                              0.0  ...   

       Away_Team_San Antonio Spurs  Away_Team_Toronto Raptors  \
0                              0.0   

In [10]:
# Add a "Signal" column which tracks the "Win_Margin" column
encoded_ten_seasons_df['Signal'] = 0.0

# When the "Win_Margin" is a greater than or equal to 0, set signal to 1
encoded_ten_seasons_df.loc[(encoded_ten_seasons_df['Win_Margin'] >= 0), 'Signal'] = 1

# When the "Win_Margin" is less than 0, set signal to -1
encoded_ten_seasons_df.loc[(encoded_ten_seasons_df['Win_Margin'] < 0), 'Signal'] = -1

# Display the dataframe
encoded_ten_seasons_df

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  \
0                          0.0                       0.0   
1                          0.0                       0.0   
2                          0.0                       0.0   
3                          0.0                       0.0   
4                          0.0                       0.0   
...                        ...                       ...   
12576                      0.0                       0.0   
12577                      0.0                       0.0   
12578                      0.0                       0.0   
12579                      0.0                       0.0   
12580                      0.0                       0.0   

       Home_Team_Brooklyn Nets  Home_Team_Charlotte Bobcats  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   
...                        ...                          ...   
12576                      0.0                          0.0   
12577                      0.0                          0.0   
12578                      0.0                          0.0   
12579                      0.0                          0.0   
12580                      0.0                          0.0   

       Home_Team_Chicago Bulls  Home_Team_Cleveland Cavaliers  \
0                          0.0                            0.0   
1                          0.0                            0.0   
2                          0.0                            0.0   
3                          0.0                            0.0   
4                          0.0                            0.0   
...                        ...                            ...   
12576                      0.0                            0.0   
12577                      0.0                            0.0   
12578                      0.0                            0.0   
12579                      0.0                            0.0   
12580                      0.0                            0.0   

       Home_Team_Dallas Mavericks  Home_Team_Denver Nuggets  \
0                             0.0                       0.0   
1                             1.0                       0.0   
2                             0.0                       0.0   
3                             0.0                       0.0   
4                             0.0                       0.0   
...                           ...                       ...   
12576                         0.0                       0.0   
12577                         0.0                       0.0   
12578                         0.0                       0.0   
12579                         0.0                       0.0   
12580                         0.0                       0.0   

       Home_Team_Detroit Pistons  Home_Team_Golden State Warriors  ...  \
0                            0.0                              0.0  ...   
1                            0.0                              0.0  ...   
2                            0.0                              0.0  ...   
3                            0.0                              0.0  ...   
4                            0.0                              1.0  ...   
...                          ...                              ...  ...   
12576                        0.0                              0.0  ...   
12577                        0.0                              0.0  ...   
12578                        0.0                              0.0  ...   
12579                        0.0                              0.0  ...   
12580                        0.0                              0.0  ...   

       Away_Team_Toronto Raptors  Away_Team_Utah Jazz  \
0                            0.0             

In [11]:
# Change the 'Signal' column to float data type
encoded_ten_seasons_df['Signal'] = encoded_ten_seasons_df['Signal'].astype(float)
encoded_ten_seasons_df.dtypes

Home_Team_Atlanta Hawks        float64
Home_Team_Boston Celtics       float64
Home_Team_Brooklyn Nets        float64
Home_Team_Charlotte Bobcats    float64
Home_Team_Chicago Bulls        float64
                                ...   
Moneyline_Home_Team            float64
Moneyline_Away_Team            float64
Game_Total_Points              float64
Win_Margin                     float64
Signal                         float64
Length: 69, dtype: object

In [12]:
# Display the amount of each signal option
encoded_ten_seasons_df['Signal'].value_counts()

 1.0    7318
-1.0    5263
Name: Signal, dtype: int64

In [13]:
# Set the y (target value) to 'Signal' column
y = encoded_ten_seasons_df['Signal']

# Set the X (features) to all columns except 'Signal'
X = encoded_ten_seasons_df.drop(columns=['Signal'])

# Display a sample of each
display(y[:5])
display(X.head())

0    1.0
1   -1.0
2   -1.0
3    1.0
4   -1.0
Name: Signal, dtype: float64

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  Home_Team_Brooklyn Nets  \
0                      0.0                       0.0                      0.0   
1                      0.0                       0.0                      0.0   
2                      0.0                       0.0                      0.0   
3                      0.0                       0.0                      0.0   
4                      0.0                       0.0                      0.0   

   Home_Team_Charlotte Bobcats  Home_Team_Chicago Bulls  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   

   Home_Team_Cleveland Cavaliers  Home_Team_Dallas Mavericks  \
0                            0.0                         0.0   
1                            0.0                         1.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0.0   

   Home_Team_Denver Nuggets  Home_Team_Detroit Pistons  \
0                       0.0                        0.0   
1                       0.0                        0.0   
2                       0.0                        0.0   
3                       0.0                        0.0   
4                       0.0                        0.0   

   Home_Team_Golden State Warriors  ...  Away_Team_San Antonio Spurs  \
0                              0.0  ...                          0.0   
1                              0.0  ...                          0.0   
2                              0.0  ...                          0.0   
3                              0.0  ...                          0.0   
4                              1.0  ...                          0.0   

   Away_Team_Toronto Raptors  Away_Team_Utah Jazz  \
0                        0.0                  0.0   
1                        0.0                  0.0   
2                        0.0                  0.0   
3                        0.0                  0.0   
4                        0.0                  0.0   

   Away_Team_Washington Wizards  Over_Under  Spread  Moneyline_Home_Team  \
0                           0.0       195.5     5.0               -210.0   
1                           0.0       190.5     3.5                185.0   
2                           0.0       186.0     1.5                205.0   
3                           0.0       197.0     7.5               -380.0   
4                           0.0       212.0     4.0                175.0   

   Moneyline_Away_Team  Game_Total_Points  Win_Margin  
0                185.0              210.0         2.0  
1               -210.0              199.0       -11.0  
2               -245.0              175.0        -1.0  
3                310.0              186.0         8.0  
4               -200.0              191.0       -19.0  

[5 rows x 68 columns]

In [14]:
# Split the data into training and testing variables
# We will NOT be scaling the data in this intial model

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    random_state=1
)

In [15]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [16]:
# Define the number of inputs to the model
inputs_1 = len(X_train.iloc[0])

# Display the amount of inputs
inputs_1

68

In [17]:
# Define the amount of outputs
outputs = 1

In [18]:
# Define the first hidden layer
hidden_1_layer_1 = ((inputs_1 + 1) // 2) + 33

# Display the amount of nodes
hidden_1_layer_1

67

In [19]:
# Define the second hidden layer
hidden_1_layer_2 = ((hidden_1_layer_1 + 1) // 2)

# Display the amount of nodes
hidden_1_layer_2

34

In [20]:
# Define the third hidden layer
hidden_1_layer_3 = ((hidden_1_layer_2 + 1) // 2)

# Display the amount of nodes
hidden_1_layer_3

17

In [21]:
# Create a Sequential mode instance
neural_network_1 = Sequential()

2023-02-14 11:42:04.666930: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [22]:
# Add the hidden layers
# First layer
neural_network_1.add(
    Dense(
        units=hidden_1_layer_1,
        input_dim=inputs_1,
        activation='ReLU'
    )
)

# Second layer
neural_network_1.add(
    Dense(
        units=hidden_1_layer_2,
        activation='ReLU'
    )
)

# Third layer
neural_network_1.add(
    Dense(
        units=hidden_1_layer_3,
        activation='ReLU'
    )
)

In [23]:
# Add the output layer
neural_network_1.add(
    Dense(
        units=outputs,
        activation='sigmoid'
    )
)

In [24]:
# Display the Sequential model summary
neural_network_1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 67)                4623      
                                                                 
 dense_1 (Dense)             (None, 34)                2312      
                                                                 
 dense_2 (Dense)             (None, 17)                595       
                                                                 
 dense_3 (Dense)             (None, 1)                 18        
                                                                 
Total params: 7,548
Trainable params: 7,548
Non-trainable params: 0
_________________________________________________________________


In [25]:
# Compile the Sequential model
neural_network_1.compile(
    loss="binary_crossentropy",
    optimizer="adam", 
    metrics=["accuracy"]
)

In [26]:
# Fit the model using 150 epochs
fit_model = neural_network_1.fit(
    X_train_scaled,
    y_train,
    epochs=150
)

Epoch 1/150
295/295 [==============================] - 5s 8ms/step - loss: -56.6957 - accuracy: 0.2007
Epoch 2/150
295/295 [==============================] - 3s 9ms/step - loss: -4754.3926 - accuracy: 0.3338
Epoch 3/150
295/295 [==============================] - 3s 8ms/step - loss: -42611.9648 - accuracy: 0.3663
Epoch 4/150
295/295 [==============================] - 3s 10ms/step - loss: -164902.5312 - accuracy: 0.3724
Epoch 5/150
295/295 [==============================] - 3s 10ms/step - loss: -433192.6875 - accuracy: 0.3785
Epoch 6/150
295/295 [==============================] - 3s 9ms/step - loss: -915309.5625 - accuracy: 0.3786
Epoch 7/150
295/295 [==============================] - 3s 9ms/step - loss: -1681633.0000 - accuracy: 0.3811
Epoch 8/150
295/295 [==============================] - 3s 9ms/step - loss: -2800338.7500 - accuracy: 0.3828
Epoch 9/150
295/295 [==============================] - 3s 11ms/step - loss: -4341739.0000 - accuracy: 0.3818
Epoch 10/150
295/295 [================

In [27]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = neural_network_1.evaluate(
    X_test_scaled,
    y_test, 
    verbose=2
)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

99/99 - 1s - loss: -5.3915e+10 - accuracy: 0.3830 - 1s/epoch - 13ms/step
Loss: -53914894336.0, Accuracy: 0.3830260634422302


## Report:
With 10 seasons of learning data scaled, using ReLU, Sigmoid, and 150 epochs, we are receiving about a 38% accuracy. This is very close to our goal of 40% accuracy. We hypothesize that the result will be better with less data, and we will try this with the more recent 5 seasons of data as well.

## Part 2: Neural Network with Deep Learning for 5 Seasons

In [28]:
# Drop the 'Date' column from the dataframe
five_seasons_nn = df_combined_2016_2021.drop(columns=['Date'])

# Display the dataframe
five_seasons_nn

Home_Team          Away_Team  Over_Under  Spread  \
0        Cleveland Cavaliers    New York Knicks       205.5     9.5   
1     Portland Trail Blazers          Utah Jazz       194.0     6.0   
2      Golden State Warriors  San Antonio Spurs       210.5     8.0   
3              Orlando Magic         Miami Heat       202.0     4.0   
4             Indiana Pacers   Dallas Mavericks       207.5     5.0   
...                      ...                ...         ...     ...   
6242            Phoenix Suns    Milwaukee Bucks       221.0     5.5   
6243         Milwaukee Bucks       Phoenix Suns       222.0     3.5   
6244         Milwaukee Bucks       Phoenix Suns       220.5     3.5   
6245            Phoenix Suns    Milwaukee Bucks       218.5     4.0   
6246         Milwaukee Bucks       Phoenix Suns       223.0     5.0   

      Moneyline_Home_Team  Moneyline_Away_Team  Game_Total_Points  Win_Margin  
0                  -450.0                355.0              205.0        29.0  
1                  -235.0                195.0              217.0         9.0  
2                  -400.0                325.0              229.0       -29.0  
3                  -155.0                135.0              204.0       -12.0  
4                  -215.0                185.0              251.0         9.0  
...                   ...                  ...                ...         ...  
6242               -180.0                160.0              226.0        10.0  
6243               -210.0                180.0              220.0        20.0  
6244               -190.0                170.0              212.0         6.0  
6245               -170.0                150.0              242.0        -4.0  
6246               -190.0                170.0              203.0         7.0  

[6247 rows x 8 columns]

In [29]:
# Review the data types
five_seasons_nn.dtypes

Home_Team               object
Away_Team               object
Over_Under             float64
Spread                 float64
Moneyline_Home_Team    float64
Moneyline_Away_Team    float64
Game_Total_Points      float64
Win_Margin             float64
dtype: object

In [30]:
# Create a list of the categorical variables
categorical_variables_2 = list(
    five_seasons_nn.dtypes[five_seasons_nn.dtypes == 'object'].index
)

# Display the list
categorical_variables_2

['Home_Team', 'Away_Team']

In [31]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(
    sparse = False
)

In [32]:
# Encode the categorical variables into numerical format
encoded_data_2 = enc.fit_transform(
    five_seasons_nn[categorical_variables_2]
)

In [33]:
# Create a dataframe with the encoded data
encoded_five_seasons_df = pd.DataFrame(
    encoded_data_2,
    columns = enc.get_feature_names_out(
        categorical_variables_2
    )
)

# Display the dataframe
encoded_five_seasons_df

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  \
0                         0.0                       0.0   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                       0.0   
...                       ...                       ...   
6242                      0.0                       0.0   
6243                      0.0                       0.0   
6244                      0.0                       0.0   
6245                      0.0                       0.0   
6246                      0.0                       0.0   

      Home_Team_Brooklyn Nets  Home_Team_Charlotte Bobcats  \
0                         0.0                          0.0   
1                         0.0                          0.0   
2                         0.0                          0.0   
3                         0.0                          0.0   
4                         0.0                          0.0   
...                       ...                          ...   
6242                      0.0                          0.0   
6243                      0.0                          0.0   
6244                      0.0                          0.0   
6245                      0.0                          0.0   
6246                      0.0                          0.0   

      Home_Team_Chicago Bulls  Home_Team_Cleveland Cavaliers  \
0                         0.0                            1.0   
1                         0.0                            0.0   
2                         0.0                            0.0   
3                         0.0                            0.0   
4                         0.0                            0.0   
...                       ...                            ...   
6242                      0.0                            0.0   
6243                      0.0                            0.0   
6244                      0.0                            0.0   
6245                      0.0                            0.0   
6246                      0.0                            0.0   

      Home_Team_Dallas Mavericks  Home_Team_Denver Nuggets  \
0                            0.0                       0.0   
1                            0.0                       0.0   
2                            0.0                       0.0   
3                            0.0                       0.0   
4                            0.0                       0.0   
...                          ...                       ...   
6242                         0.0                       0.0   
6243                         0.0                       0.0   
6244                         0.0                       0.0   
6245                         0.0                       0.0   
6246                         0.0                       0.0   

      Home_Team_Detroit Pistons  Home_Team_Golden State Warriors  ...  \
0                           0.0                              0.0  ...   
1                           0.0                              0.0  ...   
2                           0.0                              1.0  ...   
3                           0.0                              0.0  ...   
4                           0.0                              0.0  ...   
...                         ...                              ...  ...   
6242                        0.0                              0.0  ...   
6243                        0.0                              0.0  ...   
6244                        0.0                              0.0  ...   
6245                        0.0                              0.0  ...   
6246                        0.0                              0.0  ...   

      Away_Team_Oklahoma City Thunder  Away_Team_Orlando Magic  \
0                                 0.0                      0.0   
1                            

In [34]:
# Create a new variable for the exisiting numerical data and reset the index
numerical_variables_2 = five_seasons_nn.drop(
    columns = categorical_variables_2
)
numerical_variables_2

Over_Under  Spread  Moneyline_Home_Team  Moneyline_Away_Team  \
0          205.5     9.5               -450.0                355.0   
1          194.0     6.0               -235.0                195.0   
2          210.5     8.0               -400.0                325.0   
3          202.0     4.0               -155.0                135.0   
4          207.5     5.0               -215.0                185.0   
...          ...     ...                  ...                  ...   
6242       221.0     5.5               -180.0                160.0   
6243       222.0     3.5               -210.0                180.0   
6244       220.5     3.5               -190.0                170.0   
6245       218.5     4.0               -170.0                150.0   
6246       223.0     5.0               -190.0                170.0   

      Game_Total_Points  Win_Margin  
0                 205.0        29.0  
1                 217.0         9.0  
2                 229.0       -29.0  
3                 204.0       -12.0  
4                 251.0         9.0  
...                 ...         ...  
6242              226.0        10.0  
6243              220.0        20.0  
6244              212.0         6.0  
6245              242.0        -4.0  
6246              203.0         7.0  

[6247 rows x 6 columns]

In [35]:
# Add the numerical data and the categorical encoded data to one dataframe
encoded_five_seasons_df = pd.concat(
    [encoded_five_seasons_df,
     numerical_variables_2],
    axis=1
)

# Display the dataframe
encoded_five_seasons_df

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  \
0                         0.0                       0.0   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                       0.0   
...                       ...                       ...   
6242                      0.0                       0.0   
6243                      0.0                       0.0   
6244                      0.0                       0.0   
6245                      0.0                       0.0   
6246                      0.0                       0.0   

      Home_Team_Brooklyn Nets  Home_Team_Charlotte Bobcats  \
0                         0.0                          0.0   
1                         0.0                          0.0   
2                         0.0                          0.0   
3                         0.0                          0.0   
4                         0.0                          0.0   
...                       ...                          ...   
6242                      0.0                          0.0   
6243                      0.0                          0.0   
6244                      0.0                          0.0   
6245                      0.0                          0.0   
6246                      0.0                          0.0   

      Home_Team_Chicago Bulls  Home_Team_Cleveland Cavaliers  \
0                         0.0                            1.0   
1                         0.0                            0.0   
2                         0.0                            0.0   
3                         0.0                            0.0   
4                         0.0                            0.0   
...                       ...                            ...   
6242                      0.0                            0.0   
6243                      0.0                            0.0   
6244                      0.0                            0.0   
6245                      0.0                            0.0   
6246                      0.0                            0.0   

      Home_Team_Dallas Mavericks  Home_Team_Denver Nuggets  \
0                            0.0                       0.0   
1                            0.0                       0.0   
2                            0.0                       0.0   
3                            0.0                       0.0   
4                            0.0                       0.0   
...                          ...                       ...   
6242                         0.0                       0.0   
6243                         0.0                       0.0   
6244                         0.0                       0.0   
6245                         0.0                       0.0   
6246                         0.0                       0.0   

      Home_Team_Detroit Pistons  Home_Team_Golden State Warriors  ...  \
0                           0.0                              0.0  ...   
1                           0.0                              0.0  ...   
2                           0.0                              1.0  ...   
3                           0.0                              0.0  ...   
4                           0.0                              0.0  ...   
...                         ...                              ...  ...   
6242                        0.0                              0.0  ...   
6243                        0.0                              0.0  ...   
6244                        0.0                              0.0  ...   
6245                        0.0                              0.0  ...   
6246                        0.0                              0.0  ...   

      Away_Team_San Antonio Spurs  Away_Team_Toronto Raptors  \
0                             0.0                        0.0   
1                             0.0

In [36]:
# Add a "Signal" column which tracks the "Win_Margin" column
encoded_five_seasons_df['Signal'] = 0.0

# When the "Win_Margin" is a greater than or equal to 0, set signal to 1
encoded_five_seasons_df.loc[(encoded_five_seasons_df['Win_Margin'] >= 0), 'Signal'] = 1

# When the "Win_Margin" is less than 0, set signal to -1
encoded_five_seasons_df.loc[(encoded_five_seasons_df['Win_Margin'] < 0), 'Signal'] = -1

# Display the dataframe
encoded_five_seasons_df

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  \
0                         0.0                       0.0   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                       0.0   
...                       ...                       ...   
6242                      0.0                       0.0   
6243                      0.0                       0.0   
6244                      0.0                       0.0   
6245                      0.0                       0.0   
6246                      0.0                       0.0   

      Home_Team_Brooklyn Nets  Home_Team_Charlotte Bobcats  \
0                         0.0                          0.0   
1                         0.0                          0.0   
2                         0.0                          0.0   
3                         0.0                          0.0   
4                         0.0                          0.0   
...                       ...                          ...   
6242                      0.0                          0.0   
6243                      0.0                          0.0   
6244                      0.0                          0.0   
6245                      0.0                          0.0   
6246                      0.0                          0.0   

      Home_Team_Chicago Bulls  Home_Team_Cleveland Cavaliers  \
0                         0.0                            1.0   
1                         0.0                            0.0   
2                         0.0                            0.0   
3                         0.0                            0.0   
4                         0.0                            0.0   
...                       ...                            ...   
6242                      0.0                            0.0   
6243                      0.0                            0.0   
6244                      0.0                            0.0   
6245                      0.0                            0.0   
6246                      0.0                            0.0   

      Home_Team_Dallas Mavericks  Home_Team_Denver Nuggets  \
0                            0.0                       0.0   
1                            0.0                       0.0   
2                            0.0                       0.0   
3                            0.0                       0.0   
4                            0.0                       0.0   
...                          ...                       ...   
6242                         0.0                       0.0   
6243                         0.0                       0.0   
6244                         0.0                       0.0   
6245                         0.0                       0.0   
6246                         0.0                       0.0   

      Home_Team_Detroit Pistons  Home_Team_Golden State Warriors  ...  \
0                           0.0                              0.0  ...   
1                           0.0                              0.0  ...   
2                           0.0                              1.0  ...   
3                           0.0                              0.0  ...   
4                           0.0                              0.0  ...   
...                         ...                              ...  ...   
6242                        0.0                              0.0  ...   
6243                        0.0                              0.0  ...   
6244                        0.0                              0.0  ...   
6245                        0.0                              0.0  ...   
6246                        0.0                              0.0  ...   

      Away_Team_Toronto Raptors  Away_Team_Utah Jazz  \
0                           0.0                  0.0   
1                           0.0                  

In [37]:
# Change the 'Signal' column to float data type
encoded_five_seasons_df['Signal'] = encoded_five_seasons_df['Signal'].astype(float)
encoded_five_seasons_df.dtypes

Home_Team_Atlanta Hawks        float64
Home_Team_Boston Celtics       float64
Home_Team_Brooklyn Nets        float64
Home_Team_Charlotte Bobcats    float64
Home_Team_Chicago Bulls        float64
                                ...   
Moneyline_Home_Team            float64
Moneyline_Away_Team            float64
Game_Total_Points              float64
Win_Margin                     float64
Signal                         float64
Length: 67, dtype: object

In [38]:
# Display the amount of each signal option
encoded_five_seasons_df['Signal'].value_counts()

 1.0    3573
-1.0    2674
Name: Signal, dtype: int64

In [39]:
# Set the y (target value) to 'Signal' column
y = encoded_five_seasons_df['Signal']

# Set the X (features) to all columns except 'Signal'
X = encoded_five_seasons_df.drop(columns=['Signal'])

# Display a sample of each
display(y[:5])
display(X.head())

0    1.0
1    1.0
2   -1.0
3   -1.0
4    1.0
Name: Signal, dtype: float64

Home_Team_Atlanta Hawks  Home_Team_Boston Celtics  Home_Team_Brooklyn Nets  \
0                      0.0                       0.0                      0.0   
1                      0.0                       0.0                      0.0   
2                      0.0                       0.0                      0.0   
3                      0.0                       0.0                      0.0   
4                      0.0                       0.0                      0.0   

   Home_Team_Charlotte Bobcats  Home_Team_Chicago Bulls  \
0                          0.0                      0.0   
1                          0.0                      0.0   
2                          0.0                      0.0   
3                          0.0                      0.0   
4                          0.0                      0.0   

   Home_Team_Cleveland Cavaliers  Home_Team_Dallas Mavericks  \
0                            1.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0.0   

   Home_Team_Denver Nuggets  Home_Team_Detroit Pistons  \
0                       0.0                        0.0   
1                       0.0                        0.0   
2                       0.0                        0.0   
3                       0.0                        0.0   
4                       0.0                        0.0   

   Home_Team_Golden State Warriors  ...  Away_Team_San Antonio Spurs  \
0                              0.0  ...                          0.0   
1                              0.0  ...                          0.0   
2                              1.0  ...                          1.0   
3                              0.0  ...                          0.0   
4                              0.0  ...                          0.0   

   Away_Team_Toronto Raptors  Away_Team_Utah Jazz  \
0                        0.0                  0.0   
1                        0.0                  1.0   
2                        0.0                  0.0   
3                        0.0                  0.0   
4                        0.0                  0.0   

   Away_Team_Washington Wizards  Over_Under  Spread  Moneyline_Home_Team  \
0                           0.0       205.5     9.5               -450.0   
1                           0.0       194.0     6.0               -235.0   
2                           0.0       210.5     8.0               -400.0   
3                           0.0       202.0     4.0               -155.0   
4                           0.0       207.5     5.0               -215.0   

   Moneyline_Away_Team  Game_Total_Points  Win_Margin  
0                355.0              205.0        29.0  
1                195.0              217.0         9.0  
2                325.0              229.0       -29.0  
3                135.0              204.0       -12.0  
4                185.0              251.0         9.0  

[5 rows x 66 columns]

In [40]:
# Split the data into training and testing variables
# We will NOT be scaling the data in this intial model

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    random_state=1
)

In [41]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [42]:
# Define the number of inputs to the model
inputs_2 = len(X_train.iloc[0])

# Display the amount of inputs
inputs_2

66

In [43]:
# Define the amount of outputs
outputs = 1

In [44]:
# Define the first hidden layer
hidden_2_layer_1 = ((inputs_2 + 1) // 2) + 33

# Display the amount of nodes
hidden_2_layer_1

66

In [45]:
# Define the second hidden layer
hidden_2_layer_2 = ((hidden_2_layer_1 + 1) // 2)

# Display the amount of nodes
hidden_2_layer_2

33

In [46]:
# Define the third hidden layer
hidden_2_layer_3 = ((hidden_2_layer_2 + 1) // 2)

# Display the amount of nodes
hidden_2_layer_3

17

In [47]:
# Create a Sequential mode instance
neural_network_2 = Sequential()

In [48]:
# Add the hidden layers
# First layer
neural_network_2.add(
    Dense(
        units=hidden_2_layer_1,
        input_dim=inputs_2,
        activation='ReLU'
    )
)

# Second layer
neural_network_2.add(
    Dense(
        units=hidden_2_layer_2,
        activation='ReLU'
    )
)

# Third layer
neural_network_2.add(
    Dense(
        units=hidden_2_layer_3,
        activation='ReLU'
    )
)


In [49]:
# Add the output layer
neural_network_2.add(
    Dense(
        units=outputs,
        activation='sigmoid'
    )
)

In [50]:
# Display the Sequential model summary
neural_network_2.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 66)                4422      
                                                                 
 dense_5 (Dense)             (None, 33)                2211      
                                                                 
 dense_6 (Dense)             (None, 17)                578       
                                                                 
 dense_7 (Dense)             (None, 1)                 18        
                                                                 
Total params: 7,229
Trainable params: 7,229
Non-trainable params: 0
_________________________________________________________________


In [51]:
# Compile the Sequential model
neural_network_2.compile(
    loss="binary_crossentropy",
    optimizer="adam", 
    metrics=["accuracy"]
)

In [52]:
# Fit the model using 150 epochs
fit_model = neural_network_2.fit(
    X_train_scaled,
    y_train,
    epochs=150
)

Epoch 1/150
147/147 [==============================] - 4s 11ms/step - loss: -1.6688 - accuracy: 0.0862
Epoch 2/150
147/147 [==============================] - 2s 12ms/step - loss: -155.2785 - accuracy: 0.2493
Epoch 3/150
147/147 [==============================] - 2s 11ms/step - loss: -2201.1477 - accuracy: 0.3153
Epoch 4/150
147/147 [==============================] - 1s 9ms/step - loss: -10883.9141 - accuracy: 0.3358
Epoch 5/150
147/147 [==============================] - 1s 9ms/step - loss: -33120.0820 - accuracy: 0.3571
Epoch 6/150
147/147 [==============================] - 2s 14ms/step - loss: -76622.8438 - accuracy: 0.3524
Epoch 7/150
147/147 [==============================] - 2s 10ms/step - loss: -150854.0312 - accuracy: 0.3665
Epoch 8/150
147/147 [==============================] - 1s 9ms/step - loss: -264041.5625 - accuracy: 0.3637
Epoch 9/150
147/147 [==============================] - 1s 8ms/step - loss: -426997.3438 - accuracy: 0.3665
Epoch 10/150
147/147 [=======================

In [53]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = neural_network_2.evaluate(
    X_test_scaled,
    y_test, 
    verbose=2
)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

49/49 - 1s - loss: -5.8007e+09 - accuracy: 0.3534 - 696ms/epoch - 14ms/step
Loss: -5800715264.0, Accuracy: 0.3533930778503418


## Report:
With five seasons of learning data scaled, using ReLU, Sigmoid, 150 epochs, we are receiving about a 35% accuracy. This is very close to our goal of 40% accuracy, however it's not as good as our 10 seasons of data. We had hypothesized that the result would be better with less data, and it looks like we were wrong. Moving forward, we plan to adjust the amount of hidden layers and the activation type.